Second Set of Collected Bug Fixes and Maintenance Updates for 29 August 2024 Stable Release

akohlmey opened the pull request:
**Summary**

This pull request collects bugfixes backported from the develop branch to ultimately become the second update to the latest stable release.

**Author(s)**

Various, see individual commits.

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

N/A

**Implementation Notes**

The following individual changes are included:
- bugfixes for neb and neb/spin
- additional functions in linalg to support current versions of the QUIP library
- avoid hangs on errors in compute stress/mop
- bugfix in `Domain::remap_all()` to avoid lost atoms in shearing with fix deform
- fix indexing bug in dump vtk that would lead to ignoring the first 5 custom properties
- bugfix for dump netcdf when simulating systems with more than 2 billion atoms
- avoid hangs in KIM package when errors are only present on one MPI rank
- do not support contributions from bonded interactions with compute stress/mop in parallel
- various KOKKOS bugfixes
- fix a few minor issues in LAMMPS-GUI
- avoid segfault in msm/dielectric
- avoid uninitialized data access in dump styles with enabled sorting
- enable deprecated API for H5MD package
- bugfix for setting extvector for fixes in fix vector
- fixes for the Colvars library from PR #4414 and PR #4416
- use byref() instead of pointer() in python interface
- small updates for force-style unit tests to avoid cryptic crashes
- fix bugs in compute stress/cartesian
- fix bugs in fix reaxff/species
- fix normalization issue in compute rdf when the first type of a type pair is larger than the second
- fix issue when trimming neighbor lists where the trimmed list request does not have a custom cutoff
- fix issue when requesting neighbor lists with custom cutoff where neighbor lists would be re-used even when their minimal cutoff is smaller than the custom cutoff
- remove confusing warning from LAMMPS command processing
- fix segfault when using molecule files with multiple molecule IDs
- do not define __INTEL_COMPILER to __INTEL_LLVM_COMPILER to avoid problems with codepaths that are different for both compilers. Instead check for either in tests.
- backport bugfixes to the OPENMP version of pair styles hbond/dreiding/lj and hybond/dreiding/morse

**Post Submission Checklist**

- [x] The feature or features in this pull request is complete
- [x] Licensing information is complete
- [ ] Corresponding author information is complete
- [x] The source code follows the LAMMPS formatting guidelines
- [ ] Suitable new documentation files and/or updates to the existing docs are included
- [ ] The added/updated documentation is integrated and tested with the documentation build system
- [x] The feature has been verified to work with the conventional build system
- [x] The feature has been verified to work with the CMake based build system
- [ ] Suitable tests have been added to the unittest tree.
- [ ] A package specific README file has been included or updated
- [ ] One or more example input decks are included

akohlmey committed:
step version number for LAMMPS-GUI

akohlmey committed:
add LAPACK functions for matrix inversion via Cholesky decomposition

akohlmey committed:
flag branch as maintenance version again

akohlmey committed:
enable deprecated APIs for HDF5

akohlmey committed:
Backport KOKKOS bugfixes from PR #4346

akohlmey committed:
add missing declaration

akohlmey committed:
backport bugfix from PR #4296

stanmoore1 committed:
Must use alias instead of shallow copy for Kokkos hash

akohlmey committed:
avoid segfault in msm/dielectric

akohlmey committed:
change tutorial download URL to tutorial website

akohlmey committed:
remove empirical filter to remove outliers from corrupted data

akohlmey committed:
move cursor to end of log buffer before inserting new text

akohlmey committed:
small cleanup and modernization

akohlmey committed:
contributions from bonded interactions is broken when running in parallel

see https://matsci.org/t/missing-bond-contributions-from-compute-stress-mop/58455
for details.

akohlmey committed:
apply clang-format, fix other minor formatting issues, use error->one()

ndtrung81 committed:
fixed a bug with three_ilist when used with pair hybrid

stanmoore1 committed:
Fix bug when a proc has no atoms/neighbors (existing issue)

stanmoore1 committed:
Fix another (already existing) bug when a proc has no atoms

akohlmey committed:
synchronize neb/spin with bugfixes for neb

akohlmey committed:
do not free a NULL communicator

akohlmey committed:
update linalg for ML-QUIP

akohlmey committed:
correct error message

stanmoore1 commented:
@akohlmey can you include b74096b2c4578979aed8ee37461031290d3d71d1 and 0ad6babead44c48bfaeea89647495d8a95bdfe32 which are bugfixes?

ndtrung81 commented:
@akohlmey I found and fixed a memory leak issue with `fix charge/regulation` as reported by @FitzSW
```
diff --git a/src/MC/fix_charge_regulation.cpp b/src/MC/fix_charge_regulation.cpp
index ddf14f6804..f5be7a041b 100644
--- a/src/MC/fix_charge_regulation.cpp
+++ b/src/MC/fix_charge_regulation.cpp
@@ -174,9 +174,9 @@ FixChargeRegulation::~FixChargeRegulation() {
     neighbor->exclusion_group_group_delete(exclusion_group, igroupall);
   }
 
-  if (groupstrings) {
+  if (ngroups > 0) {
     for (int i = 0; i < ngroups; ++i) delete[] groupstrings[i];
-    memory->destroy(groupstrings);
+    memory->sfree(groupstrings);
   }
 }
```

akohlmey commented:
> @akohlmey can you include [b74096b](https://github.com/lammps/lammps/commit/b74096b2c4578979aed8ee37461031290d3d71d1) and [0ad6bab](https://github.com/lammps/lammps/commit/0ad6babead44c48bfaeea89647495d8a95bdfe32) which are bugfixes?

Done

akohlmey committed:
avoid memory leak

akohlmey commented:
> @akohlmey I found and fixed a memory leak issue with `fix charge/regulation` as reported by @FitzSW

Thanks. Applied to "maintenance" but also the "collection PR" so it will be in "develop" as well.

akohlmey committed:
avoid crashes with 180 degree angles

akohlmey committed:
avoid crashes with 180 degree angles

akohlmey committed:
Merge branch 'maintenance' of https://github.com/lammps/lammps into maintenance

# Conflicts:
#	src/MANYBODY/pair_sw_angle_table.cpp

giacomofiorin commented:
Hi @akohlmey! May I submit a PR for Colvars bugfixes onto the `maintenance` branch directly? Or would you rather that I target `develop` and isolate a commit for you to cherry-pick here? Thank you!!

akohlmey commented:
> Hi @akohlmey! May I submit a PR for Colvars bugfixes onto the `maintenance` branch directly? Or would you rather that I target `develop` and isolate a commit for you to cherry-pick here? Thank you!!

Both are acceptable, but submitting two pull requests is probably safer.

akohlmey commented:
@giacomofiorin come to think of it, you may have permissions to add to the maintenance branch directly and just push and only need to submit a pull request for the develop branch.

akohlmey committed:
move citeme call(s) to places where labelmaps are used

lichanghao committed:
modified dump_image.cpp, line 1100, fixed an indexing bug causing rendering error

akohlmey committed:
avoid uninitialized data access

abkein committed:
Marked multitype's default constructor as noexcept, b/c it is

jtclemm committed:
Removing extra x-lambda conversion

akohlmey committed:
backport KOKKOS bugfixes from PR #4399

stanmoore1 committed:
Fix more GPU data movement issues with fix langevin/kk and gjf option

stanmoore1 committed:
Fix another issue with gjf flag

stanmoore1 committed:
Fix more issues

akohlmey committed:
small LAMMPS-GUI flatpak build updates

akohlmey committed:
fix incorrect extvector setting bug for using fix vector with fixes

akohlmey committed:
use safe version of Domain::minimum_image() for large image flag counts.

jhenin committed:
Fixes for the Colvars library

- 759 min_image fix; Saves long runs from crashes;
  https://github.com/Colvars/colvars/pull/759 (@PolyachenkoYA)

- 728 Fix undefined behavior when getting the current working directory from std::filesystem
  https://github.com/Colvars/colvars/pull/728 (@giacomofiorin)

- 724 Fix gradients and metric functions of distanceDir
  https://github.com/Colvars/colvars/pull/724 (@giacomofiorin)

- 715 Add missing rotation in orientation component
  https://github.com/Colvars/colvars/pull/715 (@giacomofiorin)

- 713 fix: try to solve #87 for non-scala components
  https://github.com/Colvars/colvars/pull/713 (@HanatoK)

- 706 BUGFIX for Segmentation fault in colvarbias_meta::calc_energy() with useGrids off
  https://github.com/Colvars/colvars/pull/706 (@alphataubio)

- 701 Do not try accessing LAMMPS proxy object before allocating it
  https://github.com/Colvars/colvars/pull/701 (@giacomofiorin)

Authors: @alphataubio, @giacomofiorin, @HanatoK, @PolyachenkoYA

giacomofiorin committed:
Set update version to reflect upcoming LAMMPS release "29 Aug 2024 update 2"

akohlmey committed:
Merge pull request #4414 from Colvars/colvars-fixes-2024

Fixes for the Colvars library

akohlmey committed:
flag as update 2

giacomofiorin committed:
Add cherry-picked bugfix extracted from Colvars feature branch

akohlmey committed:
Merge pull request #4416 from lammps/colvars-fixes-2024

Small fix to Colvars library (1-step offset in definition of total force)

akohlmey committed:
cosmetic

akohlmey committed:
make format style variable tests unambiguous for rounding issues

stanmoore1 committed:
Fix memory corruption in comm_tiled_kokkos

akohlmey committed:
work around bug in FindVTK.cmake shipped with VTK 9.1 (and possibly other versions)

akohlmey committed:
make taper warnings consistent between Kokkos and CPU and avoid multiple outputs

akohlmey committed:
highlight error and warning messages in output window

akohlmey committed:
suppress bogus compiler warnings about overflowing buffer size in memcpy()

akohlmey committed:
memcpymask is a constant and thus should be uppercase

akohlmey committed:
logo images for all 8 currently supported tutorials

akohlmey committed:
add code paths for all 8 tutorial wizards. only descriptive text is missing.

akohlmey committed:
sync LAMMPS-GUI howto with upstream as it describes the same GUI version

akohlmey committed:
add overloads for automatic string style conversions

akohlmey committed:
show download progress in status line and progress bar

akohlmey committed:
combine the first two and the second two wizard pages into one page each

akohlmey committed:
display panel with number or warnings and lines in output window, button for jumping to next warning

akohlmey committed:
update tutorial info texts from paper

akohlmey committed:
compatibility with Qt 5.12 on Ubuntu 20.04LTS

akohlmey committed:
improve layout of warnings panel

akohlmey committed:
add keyboard shortcut and context menu entry for jump to next warning

akohlmey committed:
add check whether libcurl support was compiled into LAMMPS and geturl is functional

akohlmey committed:
check if a downloaded (solution) file is a symlink placeholder and put a copy of the original in its place

akohlmey committed:
display a valid URL for the download location

akohlmey committed:
include WHAM software in LAMMPS-GUI compilation

akohlmey committed:
correct CMake script code for including WHAM executables in macOS app-bundle

akohlmey committed:
document inclusion of WHAM

akohlmey committed:
add patch for customization of wham to support LAMMPS units

akohlmey committed:
update patch for WHAM

riseofdata committed:
Single process errors in pair style kim reported using error->one

Errors during a KIM compute operation that occur on a single processor were reported using error->all, which causes LAMMPS to hang when running in parallel with more than one processor. This has been fixed by replacing error->all with error->one for those cases.

akohlmey committed:
fix indexing bug in dump vtk that would ignore the first 5 custom properties

akohlmey committed:
fix typo

akohlmey committed:
MPI bugfix for dump netcdf from Paul Coffman

akohlmey committed:
use mutex to avoid race condition when accessing thermo data during run

akohlmey committed:
macOS does not like forward declarations for standard C++ classes

rbberger committed:
add missing update to invoked_bonds in ComputeReaxFFAtomKokkos

rbberger commented:
@akohlmey this should probably also include a22c58cd47b2e5d80196e8a5e222e5a3bc7c9821

akohlmey committed:
add dummy variant of Run WHAM dialog

akohlmey committed:
update patch for WHAM code

akohlmey committed:
increment bugfix for "inputs local"

akohlmey committed:
we need tighter checks, also on the smallest pairwise cutoff, before we can re-use a default neighbor list for an occasional list with an explicit cutoff

akohlmey committed:
display error messages with fixed width font

akohlmey committed:
prevent the neighbor list re-ordering from getting stuck

akohlmey committed:
must not set to unique if request is for skip list. only check for smallest pair cutoff.

akohlmey committed:
add option to restart the LAMMPS instance to "Run" menu

stanmoore1 committed:
Fix more issues in fix langevin/kk pointed out by @ndtrung81

Co-authored-by: Trung Nguyen <ndactrung@gmail.com>
Co-authored-by: Stan Moore <stanmoore1@gmail.com> 

stanmoore1 committed:
Fix small memory leak, add debug RNG

stanmoore1 committed:
Fix more issues in Kokkos fix langevin gjf option

stanmoore1 committed:
Need local capture for lamdas on GPUs

stanmoore1 reviewed (approved):
Need to port the last few commits from https://github.com/lammps/lammps/pull/4399 to here, then this is ready to merge from my POV.

akohlmey committed:
fix fmt::format() missing argument bugs

akohlmey committed:
add test whether drude particle was found locally and stop with error when not

this handles two cases:
1 the actual drude particle is not in the sub-domain
2 the drude particle was never assigned by fix drude

akohlmey committed:
enable and apply clang-format

akohlmey committed:
add check to fix drude to detect if core atoms without drude atom id exist

akohlmey committed:
correct help text

akohlmey committed:
fix for old error class

stanmoore1 committed:
Fix bug in compute stress/cartesian density profile with periodic boundary conditions

akohlmey committed:
set lmp pointer to NULL after delete to avoid using it or deleting it a second time

stanmoore1 committed:
Fix bug in compute_stress_cartesian with periodic boundary conditions

akohlmey committed:
remove sometimes misleading and often confusing warning

akohlmey committed:
bugfix from @jtclem for molecule files with multiple molecule IDs

akohlmey committed:
do not define __INTEL_COMPILER to __INTEL_LLVM_COMPILER instead test for either

This shortcut will create problems for features that do not exist for
the Intel LLVM based compiler.

akohlmey committed:
fix small memory leak

akohlmey committed:
backport fixes to hbond/dreiding pair styles from upstream, add unit tests

akohlmey committed:
refactor catching exceptions during LAMMPS initialization

this will avoid the nasty segfaults with "cannot have multiple stdout capturers"
instead it will catch and display any exception thrown during init.

akohlmey committed:
use byref() instead of pointer()

akohlmey committed:
backport fixes to fix reaxff/species

stanmoore1 committed:
Fix rare bug in KOKKOS, manifest when load balancing on GPUs and exchange comm is on host

akohlmey committed:
make certain that Contact::varflag is initialized

stanmoore1 committed:
Fix view with wrong label

akohlmey committed:
correctly compute cutneighmin when multiple requests with different cutoff exist

akohlmey committed:
move update to cutneighmin from neighbor lists requests with explicit cutoff to a better location

akohlmey committed:
fix copy and paste bug

akohlmey committed:
fix bug in angle style cosine/delta

rbberger commented:
~Please pull in 4d453a65e6728760d3eeb0e88ea2ad8a49d790e1.~ nevermind, that function doesn't exist in that release yet.

akohlmey committed:
fix typo

akohlmey committed:
move precomputation for factor_sqrt to individual Pair::compute() functions

the special_lj values may be changed for individual hybrid sub-styles
with pair_modify pair special. thus the factor_sqrt[] array may have
incorrect values when computed during Pair::init_style().

rbberger committed:
adjust BondStyle:class2 epsilon for aarch64 and ppc64le builds

akohlmey committed:
backport indexing bugfix from upstream

akohlmey committed:
add test to detect if a potential file was incorrectly used with lgvdw yes

akohlmey committed:
bugfix from Trung to avoid problems with read_dump when not all MPI ranks have atoms

akohlmey committed:
only trim if neighbor list request has a custom cutoff

akohlmey committed:
use pairwise cutoff for trimming unless a custom cutoff if given

akohlmey committed:
use different method to enforce the C locale in LAMMPS GUI

akohlmey committed:
reverse normalization between type pairs if the types were swapped on input

akohlmey committed:
rewrite incorrect statement

akohlmey committed:
error out when trying unsupported pimd method with multiple processors per bead

akohlmey committed:
set stores_ids flag for some fixes as suggested by @sjplimp

akohlmey committed:
properly tally the SYNC time contributions to the ALL time entry

this way its contribution is not double counted when determining
the "Other" time data.

akohlmey committed:
fix bug missing a read_int() when reading general triclinic rotation matrix

akohlmey committed:
update to WHAM version 2.1.0

akohlmey committed:
temporarily comment out incomplete run WHAM and plot results dialog for release
